File: scripts/algorithms/N/Next Greater Element II/Next Greater Element II.js
// Runtime: 218 ms (Top 26.53%) | Memory: 48.7 MB (Top 35.44%)
var nextGreaterElements = function(nums) {
    const len = nums.length;
    const ans = new Array(len).fill(-1);
    const stack = [];
    for(let i = 0; i < len; i++) {
        if(i == 0) stack.push([i, nums[i]]);
        else {
            while(stack.length && stack.at(-1)[1] < nums[i]) {
                ans[stack.pop()[0]] = nums[i];
            }
            stack.push([i, nums[i]]);
        }
    }
    for(let num of nums) {
        while(stack.length && stack.at(-1)[1] < num) {
            ans[stack.pop()[0]] = num;
        }
    }
    return ans;
};

File: scripts/algorithms/N/Next Greater Element II/Next Greater Element II.py
// Runtime: 151 ms (Top 99.29%) | Memory: 19.30 MB (Top 5.84%)

class Solution:
    def nextGreaterElements(self, nums: List[int]) -> List[int]:
        st = []
        n = len(nums)
        ans = [-1] * n
        for i in range(2*n-1, -1, -1):
            while st and st[-1] <= nums[i%n]:
                st.pop()
            if st and i < n:
                ans[i] = st[-1]
            st.append(nums[i%n])
        return ans
